bob-backend/src/core.rs

```diff
@@ -131,10 +134,66 @@ pub trait MetricsProducer: Send + Sync {
     }
 }
 
+const ERROR_LOG_INTERVAL: u64 = 5000;
+
+#[derive(Hash, PartialEq, Eq, Debug)]
+enum BackendErrorAction {
+    PUT(String, String),
+}
+
+#[derive(Debug)]
+struct BackendErrorLogger {
+    errors: HashMap<BackendErrorAction, u64>,
+    interval: CDuration,
+    last_timestamp: CInstant,
+}
+
+impl BackendErrorLogger {
+    fn new() -> BackendErrorLogger {
+        BackendErrorLogger {
+            errors: HashMap::new(),
+            interval: CDuration::from_millis(ERROR_LOG_INTERVAL),
+            last_timestamp: CInstant::now(),
+        }
+    }
+
+    fn report_error(logger: Arc<Mutex<BackendErrorLogger>>, action: BackendErrorAction) {
```

Comment:
All methods in this type should work directly with inner fields. `Arc` and `Mutex` are the things, that are created in the place of use when needed

Reply:
Or you can take another approach: declare inner struct with HashMap and Timestamp and protect it with Mutex inside this struct

Reply:
One problem: Mutex potentially can be under a high contention. Try to use Mutex from https://crates.io/crates/parking_lot crate. It implements spinning strategy, that should highly improve performance in our scenario